### PR TITLE
fix(context): join all system messages instead of taking only the first (#390)

### LINKED ISSUE #390: Every system message after the first is silently discarded
## Problem

Every `system` message after the first is silently discarded before token counting and generation.

`ContextManager.makeSession` strips **all** system messages out of the conversation, and
`buildInstructions` then copies only the **first** one back into the transcript. Everything in
between is gone, with no warning, no error, and no token accounting.

## Where

`Sources/ContextManager.swift:35`, `Sources/ContextManager.swift:127-130` (commit b39659abe6b627e95454c8144d8fdeff9b3f50a3)

The same path serves Responses `developer` messages (`Sources/Core/ResponsesModels.swift:140-158`)
and CLI `--messages` conversations combined with `-s`.

## Evidence

Token accounting proves the message never reaches the model. Two requests to `/v1/chat/completions`
differing only by an extra ~400-word system message:

```
[system "Be concise.",                          user "Say OK."]  -> prompt_tokens 28
[system "Be concise.", system <400 words>,      user "Say OK."]  -> prompt_tokens 28
```

A 400-word instruction contributes exactly zero tokens.

## Impact

Multi-system-message conversations are ordinary: SDKs and frameworks routinely emit a base system
prompt plus a task-specific one, and LangChain/OpenAI-style agents stack them. apfel accepts the
request, returns HTTP 200, and answers while ignoring instructions the caller believes are in
force. This is silent, not loud - the caller has no way to detect it short of counting tokens.

## Reproduction

With `apfel --serve --port 11434` running:

```sh
python3 - <<'PY'
import json, urllib.request
def post(msgs):
    req = urllib.request.Request('http://127.0.0.1:11434/v1/chat/completions',
        data=json.dumps({'model':'apple-foundationmodel','messages':msgs,'max_tokens':8}).encode(),
        headers={'Content-Type':'application/json'})
    with urllib.request.urlopen(req, timeout=60) as r:
        return json.load(r)['usage']['prompt_tokens']
base  = [{'role':'system','content':'Be concise.'}, {'role':'user','content':'Say OK.'}]
extra = [{'role':'system','content':'Be concise.'},
         {'role':'system','content':'Additional instruction. ' * 100},
         {'role':'user','content':'Say OK.'}]
print('one system :', post(base))
print('two systems:', post(extra))
PY
```

Observed: `one system : 28` and `two systems: 28`.

## Concrete fix

Join every system message into the instructions block instead of taking only the first.

```diff
--- a/Sources/ContextManager.swift
+++ b/Sources/ContextManager.swift
@@ -125,8 +125,15 @@
         // System prompt
-        if let sys = messages.first(where: { $0.role == "system" })?.textContent {
-            parts.append(sys)
+        // Every system message must survive, not just the first: makeSession
+        // filters them all out of the conversation (line 35), so any system
+        // message not copied here is silently dropped before token counting
+        // and generation.
+        let systemParts = messages
+            .filter { $0.role == "system" }
+            .compactMap { $0.textContent }
+            .filter { !$0.isEmpty }
+        if !systemParts.isEmpty {
+            parts.append(systemParts.joined(separator: "\n\n"))
         }
```

## Acceptance criteria

- [ ] Two system messages produce a strictly larger `prompt_tokens` than one
- [ ] The content of every system message is present in the built instructions, in order
- [ ] A conversation with zero system messages behaves exactly as today
- [ ] Responses `developer` messages and CLI `--messages` + `-s` get the same treatment
- [ ] If a future design decides only one system message is supported, the extra ones are rejected with a 400 rather than dropped

## Tests to add

- [ ] `Tests/apfelTests/ContextManagerTests.swift` - `testAllSystemMessagesReachInstructions()` asserts both system contents appear in the built instructions
- [ ] `Tests/apfelTests/ContextManagerTests.swift` - `testNoSystemMessageIsUnchanged()` asserts the empty case is untouched
- [ ] `Tests/integration/server_validation_test.py` - `test_second_system_message_changes_prompt_tokens` asserts `prompt_tokens` grows when a second system message is added

---
<sub>Filed by automated review: OpenAI Codex `gpt-6-astra` (xhigh) reviewed, Claude Code verified against commit b39659abe6b627e95454c8144d8fdeff9b3f50a3 on 2026-09-06. Pass: edge-case.</sub>

### PR DESCRIPTION
**Pull request by Arthur-Ficial**
**Draft PR - needs @franzenzenhofer review and local test run before merging.**

Fixes #390.

## Root cause

`ContextManager.buildInstructions` (line 128) used `.first(where: { $0.role == "system" })` to extract the system prompt, picking up only the first system message and silently discarding every subsequent one. Meanwhile `makeSession` (line 35) filters all system messages out of the conversation array, so any system message not copied into the instructions block by `buildInstructions` vanishes entirely - no warning, no error, no token accounting.

Multi-system-message conversations are ordinary: SDKs and frameworks routinely stack a base system prompt with task-specific ones, and the Responses API `developer` role maps to `system` via `ResponsesMapper`, compounding the problem when `instructions` is also set.

## The fix

Added `OpenAIMessage.joinedSystemContent(from:)` as a public static method in `ApfelCore` (`Sources/Core/OpenAIModels.swift`). It collects every system-role message's text content (preserving order, skipping empty/nil), and joins them with `\n\n`. Returns nil when no system message carries non-empty text.

`ContextManager.buildInstructions` now calls this helper instead of `.first(where:)`, so every system message reaches the instructions block and is counted in prompt tokens.

## Test coverage

- Added `Tests/apfelTests/SystemMessageJoinTests.swift` with 8 unit tests:
  - `testSingleSystemMessageReturnsContentUnchanged`
  - `testTwoSystemMessagesJoinedWithDoubleNewline`
  - `testThreeSystemMessagesPreserveOrder`
  - `testNoSystemMessagesReturnsNil`
  - `testEmptyMessagesArrayReturnsNil`
  - `testSystemMessagesWithEmptyContentSkipped`
  - `testSystemMessageWithNilContentSkipped`
  - `testOnlyEmptySystemMessagesReturnsNil`
  - `testDeveloperMappedSystemMessagesJoined` (Responses path)
- Existing `OpenAIModelsTests` and `ResponsesModelsTests` cover the types used.

## What I verified (static only)

- The fix correctly replaces the `.first(where:)` call with a collect-all-and-join approach.
- The helper is a pure function with no concurrency implications.
- The empty/nil/single-message paths behave identically to today's code.
- `ResponsesMapper.messages(from:)` maps `developer` role to `system`, so the fix covers the Responses path automatically.
- No `@unchecked Sendable`, no data races.
- Diff limited to `Sources/Core/OpenAIModels.swift`, `Sources/ContextManager.swift`, `Tests/apfelTests/SystemMessageJoinTests.swift`, `Tests/apfelTests/main.swift`, `CHANGELOG.md` - no touches to release infrastructure, guardrails, CI, or dependencies.

## What I did NOT verify

- **Functional correctness. This needs `make test` on a Mac with Apple Intelligence.** I cannot run FoundationModels code. If the fix does not actually resolve the reported behaviour, that is on me to refine - ping me in a review comment.

## Anything that seemed suspicious

Nothing - straightforward bug. The issue is filed by the owner account (`Arthur-Ficial`), the report includes a clear reproducer, and the root cause is a simple `.first(where:)` that should be a `.filter`.

---

cc @franzenzenhofer - draft stays draft until you review, test locally, and mark ready.

_Generated by the apfel bug-solver routine._

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01WuWnEGFHqvz9qQX4DLzQRL

---
_Generated by [Claude Code](https://claude.ai/code/session_01WuWnEGFHqvz9qQX4DLzQRL)_

### CODE REVIEW

**Comment by Arthur-Ficial** [comment]
The `apfelcore-public-api` check failed, but not from anything in this PR.

The API breakage step itself passed cleanly ("No breaking changes detected in ApfelCore"). The failure is in the subsequent "Install Python test dependencies" step: `pip3 install --break-system-packages pytest httpx` fails because the runner's brew-managed `typing_extensions` 4.15.0 has no RECORD file, so pip refuses to uninstall it to upgrade to 4.16.0. This is a runner environment regression - the same step passed on main (run 31016768088, Aug 5) on an older runner image.

The two checks that exercise this PR's code both passed:
- `build-and-test`: success (build + 1042 unit tests + model-free integration)
- `changelog-gate`: success

I lack the means to re-run the job (CI workflow changes are forbidden for this routine). The fix would be pinning `typing_extensions` or using `--ignore-installed` in the CI workflow, but that's `.github/workflows/` territory - flagging for Franz.

cc @franzenzenhofer

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Arthur-Ficial — COMMENTED** [review]
## First-pass triage review — automated daily pass, 2026-09-07

Reviewing against the golden goal (UNIX tool + OpenAI-compatible server + chat, 100% on-device) and the PR process in `CLAUDE.md`. This is a **first-pass safety net, not an approval** — I have not merged, approved, or pushed anything.

**Classification:** Code: network/parsing/auth · `+107/-2` across 5 files · targets #390

### Process checks (`CLAUDE.md` §7)
- ✅ Code change ships with tests.
- ✅ `CHANGELOG.md` touched (the `changelog-gate` requirement for any `Sources/**` change).

### CI
Red on: `apfelcore-public-api` — but **this is not your diff.**

Every red check across all 61 open PRs in this repo currently fails at the *same* step, `Install Python test dependencies`:

```
× Cannot uninstall typing_extensions 4.15.0
╰─> no RECORD file was found for typing_extensions (installed by brew)
```

It is a flaky runner-image issue, not deterministic — the same commit can pass on a re-run. **Re-running the job is usually enough.** Root cause and fix options are in #438.

### ⚠️ 2 competing candidates for #390

The auto-fix routine opened more than one PR for this issue. Only one should land; the rest should be closed.

| PR | size | tests | CI | 
| --- | --- | --- | --- |
| **#400 (this one)** | `+107/-2` | yes | 🔴 pip flake only |
| #403 | `+80/-2` | yes | 🟢 green |

Because the CI signal is unreliable right now, **do not use green vs red as the tiebreaker** between these. The #462 group is the cautionary precedent: all three candidates were plausible and CI-green, and all three were wrong (the guard turned a crash into silent schema truncation). Compare the diffs on their merits and run the tests locally.

**Status: draft, awaiting @franzenzenhofer.** Franz merges, Franz releases — always.

Cheers, Arthur

**Comment by Arthur-Ficial** [comment]
Closing: #390 is fixed on `main` in e359a2c.

**Not merged: a near-identical candidate was taken instead** (#403).

The routine opened several candidates for this issue with substantively the same change. Only one can land; this one is closed as a duplicate, not as wrong. Where this PR's tests were better than the merged one's, they were carried across.

The landed change is verified, not just plausible: every fix in this sweep was built and re-run against the original reproduction before being committed.

_Closed as part of the 2026-09-07 backlog sweep: 29 issues fixed, 61 candidate PRs adjudicated._
